korangar/src/loaders/animation/mod.rs

```diff
@@ -0,0 +1,474 @@
+use std::cmp::{max, min};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use cgmath::Vector2;
+use derive_new::new;
+use korangar_interface::elements::PrototypeElement;
+use num::Zero;
+use wgpu::{Device, Extent3d, Queue, TextureDescriptor, TextureDimension, TextureFormat, TextureUsages};
+
+use super::error::LoadError;
+use crate::graphics::Texture;
+use crate::loaders::{ActionLoader, Actions, AnimationState, Sprite, SpriteLoader};
+use crate::{Color, EntityType};
+
+// TODO: use cache later, the memory will be increase with this hashmap, until
+// the program is out of memory.
+#[derive(new)]
+pub struct AnimationLoader {
+    device: Arc<Device>,
+    queue: Arc<Queue>,
+    #[new(default)]
+    // The string will be type of entity
+    // 0_{body_id}_{head_id}
+    // 1_{monster_id}
+    // 2_{npc_id}
+    cache: HashMap<String, AnimationData>,
+}
+
+impl AnimationLoader {
+    pub fn load(
+        &mut self,
+        sprite_loader: &mut SpriteLoader,
+        action_loader: &mut ActionLoader,
+        entity_filename: Vec<String>,
+        entity_type: EntityType,
+    ) -> Result<AnimationData, LoadError> {
+        // Create animation pair with sprite and action
+        let vec: Vec<AnimationPair> = entity_filename
+            .iter()
+            .map(|file_path| AnimationPair {
+                sprites: sprite_loader.get(&format!("{file_path}.spr")).unwrap(),
+                actions: action_loader.get(&format!("{file_path}.act")).unwrap(),
+            })
+            .collect();
+
+        // The sprite is stored as rgba in animation_pair.sprites.rgba_images or
+        // animation_pair.sprites.palette_images
+        // The sprite is stored as texture in animation_pair.sprites.textures
+
+        // For each animation, we collect all the framepart need to generate the frame
+        let mut animations_list: Vec<Vec<Vec<Frame>>> = Vec::new();
+        let mut animation_index: usize = 0;
+        let mut action_index: usize = 0;
+        let mut motion_index: usize = 0;
+
+        // Each animation pair has the sprites and actions, we iterate over the
+        // animation pairs
+        // Each entity has several actions and the actions is composed of several
+        // motion, in each motion contains several pictures that we try to
+        // merge.
+        for animation_pair in vec.iter() {
+            action_index = 0;
+            let mut animation_frames: Vec<Vec<Frame>> = Vec::new();
+            for action in animation_pair.actions.actions.iter() {
+                motion_index = 0;
+                let mut action_frames: Vec<Frame> = Vec::new();
+                for motion in action.motions.iter() {
+                    let mut motion_frames: Vec<Frame> = Vec::new();
+                    if motion.sprite_clip_count == 0 {
+                        continue;
+                    }
+                    for position in 0..motion.sprite_clip_count {
+                        let frame_part: FramePart;

```

Comment:
Move the declaration and initialization together.

Reply:
Fixed